### PR TITLE
feat(demos): adds a password generator demo

### PR DESCRIPTION
**Pull request by MariaAga**
**What**: Closes https://github.com/patternfly/patternfly-react/issues/6686

### CODE REVIEW

File: packages/react-core/src/demos/PasswordGenerator.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+          name="password-generator-demo--initial-password"
+          aria-label="Password input"
+          placeholder="Password"
+          iconVariant="fa-eye"
+        />
+        <Button variant="control" aria-label="search button for search input">
+          <EyeIcon />
+        </Button>
+      </InputGroup>
+      <Menu isPlain>
```
Comment (nicolethoen) [inline packages/react-core/src/demos/PasswordGenerator.md:66]: So the HTML demo that this is modeled after is meant to have the menu appear as a sort of dropdown when the user puts focus into the password input.

And usually the demos we build in react are supposed to have all interactions implemented. So the 'refresh' button should suggest a new password, and clicking the suggestion should have it auto populate the password field.

The demo should also be fully keyboard accessible - so I'd look at an existing demo - like the [search input with autocomplete](ct-demos#search-with-autocomplete) demo as an example for how using arrow keys, tab, escape, enter, etc should allow the user to interact with the suggested password

**Review by nicolethoen — CHANGES_REQUESTED** [review]

File: packages/react-core/src/demos/PasswordGenerator.md (hunk 1 above)
Comment (mcoker) [reply packages/react-core/src/demos/PasswordGenerator.md:66]: Also the show/hide (eye icon) button should toggle the text input type, showing and hiding the password in the input field. We have a login demo that shows that interaction - https://www.patternfly.org/v4/components/login-page/react/showhide-password/

File: packages/react-core/src/demos/PasswordGenerator.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    >
+      <InputGroup>
+        <TextInput
+          isRequired
+          type="password"
+          id="password-generator-demo--initial-password"
+          name="password-generator-demo--initial-password"
+          aria-label="Password input"
+          placeholder="Password"
+          iconVariant="fa-eye"
```
Comment (mcoker) [inline packages/react-core/src/demos/PasswordGenerator.md:60]: I don't think you need this?

File: packages/react-core/src/demos/PasswordGenerator.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        />
+        <Button variant="control" aria-label="search button for search input">
```
Comment (mcoker) [inline packages/react-core/src/demos/PasswordGenerator.md:62]: This should be sentence case and toggle between show/hide password. You can reference the login page show/hide password demo for that - https://www.patternfly.org/v4/components/login-page/react/showhide-password/

```suggestion
        <Button variant="control" aria-label="Show password">
```

File: packages/react-core/src/demos/PasswordGenerator.md (hunk 1 above)
Comment (mcoker) [inline packages/react-core/src/demos/PasswordGenerator.md:66]: With @nicolethoen's comments, this should also be a regular menu (not plain)

```suggestion
      <Menu>
```

**Review by mcoker — CHANGES_REQUESTED** [review]
Looking good! Left some feedback for review.

File: packages/react-core/src/demos/PasswordGenerator.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
+---
+id: Password generator
+section: demos
+---
+
+import RedoIcon from '@patternfly/react-icons/dist/esm/icons/redo-icon';
+import EyeIcon from '@patternfly/react-icons/dist/esm/icons/eye-icon';
+import HelpIcon from '@patternfly/react-icons/dist/esm/icons/help-icon';
+
+## Examples
```
Comment (tlabaj) [inline packages/react-core/src/demos/PasswordGenerator.md:10]: This should be `Demos`.

**Review by tlabaj — CHANGES_REQUESTED** [review]

**Comment by MariaAga** [comment]
Thank! Made it functional and keyboard accessible :)

**Review by mcarrano — COMMENTED** [review]
This looks good to me. @shirimordechay can you also take a look at this and see if it behaves as you expect? See the link here: https://patternfly-react-pr-6725.surge.sh/demos/password-generator

**Review by mcoker — APPROVED** [review]
LGTM!! 🤩

**Comment by shirimordechay** [comment]
Looks great! btw, the demo on https://patternfly-react-pr-6725.surge.sh/demos/password-generator didn't work for me, only in full screen.

**Comment by MariaAga** [comment]
@shirimordechay Thanks! can you share a screenshot or the console output maybe? I see it fine in all screen sizes

**Comment by shirimordechay** [comment]
@MariaAga There was no suggested password shown (when the field is active) yesterday when I checked it but It works now so I guess it's ok :) thanks!

**Review by wise-king-sullyman — CHANGES_REQUESTED** [review]
This looks really good!

I have one major change request (this component being in TS rather than JS) and a few small nits.

File: packages/react-core/src/demos/PasswordGenerator.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    window.addEventListener('keydown', handleMenuKeys);
+    window.addEventListener('click', handleClickOutside);
+    return () => {
+      window.removeEventListener('keydown', handleMenuKeys);
+      window.removeEventListener('click', handleClickOutside);
+    };
+  }, [isAutocompleteOpen, searchInputRef.current]);
+
+  const onChange = newValue => {
+    if (searchInputRef && searchInputRef.current && searchInputRef.current.contains(document.activeElement)) {
```
[the remark this answers is not in the document]
Comment (MariaAga) [reply packages/react-core/src/demos/PasswordGenerator.md:60]: I agree but when I tried to do it i get an error `invalid JSX: SyntaxError: Unexpected token`

[the remark this answers is not in the document]
Comment (wise-king-sullyman) [reply packages/react-core/src/demos/PasswordGenerator.md:60]: 🤔 Interesting, I've used it elsewhere in the repo previously. That *was* in TypeScript though and maybe TS was transpiling it to the expanded form when it generated the JS or something.

[the remark this answers is not in the document]
Comment (wise-king-sullyman) [reply packages/react-core/src/demos/PasswordGenerator.md:60]: This has been a rabbit hole through the MD parser that I've chased and I've run out of time to investigate.

Looks like optional chaining of examples/demos is off the table until this issue with the parser can be addressed so I'm going to mark this as resolved and spin up a ticket for it if we don't have one yet.

File: packages/react-core/src/demos/PasswordGenerator.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+import RedoIcon from '@patternfly/react-icons/dist/esm/icons/redo-icon';
+import EyeIcon from '@patternfly/react-icons/dist/esm/icons/eye-icon';
+import EyeSlashIcon from '@patternfly/react-icons/dist/esm/icons/eye-slash-icon';
+
+## Demos
+
+### Provide a generated password
+
+```js
```
[the remark this answers is not in the document]
Comment (MariaAga) [reply packages/react-core/src/demos/PasswordGenerator.md:14]: Should there be any code changes after changing to `ts`? I saw some demos that use `ts` but don't use types

[the remark this answers is not in the document]
Comment (wise-king-sullyman) [reply packages/react-core/src/demos/PasswordGenerator.md:14]: I think the only hard rules we have are that the component should be typed as a `React.FunctionComponent` and that anything implicitly typed as `any` should be given an explicit type.

File: packages/react-core/src/demos/PasswordGenerator.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        setIsAutocompleteOpen(false);
+        searchInputRef.current.focus();
+        // the up and down arrow keys move browser focus into the autocomplete menu
+      } else if (event.key === 'ArrowDown' || event.key === 'ArrowUp') {
+        const firstElement = autocompleteRef.current.querySelector('li > button:not(:disabled)');
+        firstElement && firstElement.focus();
+        event.preventDefault(); // by default, the up and down arrow keys scroll the window
+      }
+      // If the autocomplete is open and the browser focus is in the autocomplete menu
+      // hitting tab will close the autocomplete and but browser focus back on the search input.
```
[the remark this answers is not in the document]
Comment (MariaAga) [reply packages/react-core/src/demos/PasswordGenerator.md]: Nice catch!

File: packages/react-core/src/demos/PasswordGenerator.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+  const onSelect = e => {
+    e.stopPropagation();
+    setPassword(generatedPassword);
+    setIsAutocompleteOpen(false);
+    searchInputRef.current.focus();
+  };
+
+  const handleMenuKeys = event => {
+    if (isAutocompleteOpen && searchInputRef.current && searchInputRef.current.contains(event.target)) {
+      // the escape key closes the autocomplete menu and keeps the focus on the search input.
```
[the remark this answers is not in the document]
Comment (MariaAga) [reply packages/react-core/src/demos/PasswordGenerator.md]: Makes sense, changed it

File: packages/react-core/src/demos/PasswordGenerator.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (MariaAga) [reply packages/react-core/src/demos/PasswordGenerator.md]: Changed to ts :)

File: packages/react-core/src/demos/PasswordGenerator.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+      setIsAutocompleteOpen(false);
+      searchInputRef.current.focus();
+      // the up and down arrow keys move browser focus into the autocomplete menu
+    } else if (event.key === 'ArrowDown' || event.key === 'ArrowUp') {
+      const firstElement = autocompleteRef.current.querySelector('li > button:not(:disabled)');
+      firstElement && firstElement.focus();
+      event.preventDefault(); // by default, the up and down arrow keys scroll the window
+    }
+    // If the autocomplete is open and the browser focus is in the autocomplete menu
+    // hitting tab will close the autocomplete and FormGroup browser focus back on the search input.
```
Comment (wise-king-sullyman) [inline packages/react-core/src/demos/PasswordGenerator.md:90]:
```suggestion
    // hitting tab will close the autocomplete and put browser focus back on the search input.
```

**Review by wise-king-sullyman — COMMENTED** [review]
Looks like a word got mixed up here when you were doing the last round of changes, everything else looks great though and once that's addressed I'll be approving 🙂

File: packages/react-core/src/demos/PasswordGenerator.md (hunk 9 above)
Comment (MariaAga) [reply packages/react-core/src/demos/PasswordGenerator.md]: just a small spelling error :upside_down_face:

**Review by wise-king-sullyman — APPROVED** [review]
Looks great! Thank you! 🚀

**Review by nicolethoen — APPROVED** [review]

**Review by tlabaj — APPROVED** [review]
LGTM

**Pull request merged**